rust/sdk/src/client.rs

```diff
@@ -457,6 +458,684 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
         .flatten()
 }
 
+/// Request builder for event publishing.
+///
+/// Warning: [`Publish`] implements the [`Future`] trait, as such it can be polled.
+/// Calling _any_ [`Publish`] function after polling will result in a panic!
+pub enum Publish<'a> {
+    Initial { client: &'a Ax, request: PublishRequest },
+    Pending(BoxFuture<'a, anyhow::Result<PublishResponse>>),
+    Void,
+}
+
+impl<'a> Publish<'a> {
+    fn new(client: &'a Ax) -> Self {
+        Self::Initial {
+            client,
+            request: PublishRequest { data: vec![] },
+        }
+    }
+
+    /// Add an event.
+    ///
+    /// Subsequent calls to this function will not remove previous events.
+    ///
+    /// Example:
+    /// ```no_run
+    /// use actyx_sdk::{tags, Ax, AxOpts, service::PublishResponse};
+    /// async fn event_example() -> PublishResponse {
+    ///     let service = Ax::new(AxOpts::default()).await.unwrap();
+    ///     service
+    ///         .publish()
+    ///         .event(
+    ///             tags!("temperature", "sensor:temp-sensor1"),
+    ///             &serde_json::json!({ "temperature": 10 }),
+    ///         )
+    ///         .unwrap()
+    ///         .event(
+    ///             tags!("temperature", "sensor:temp-sensor2"),
+    ///             &serde_json::json!({ "temperature": 21 }),
+    ///         )
+    ///         .unwrap()
+    ///         .await
+    ///         .unwrap()
+    /// }
+    /// ```
+    pub fn event<E: Serialize>(mut self, tags: TagSet, event: &E) -> Result<Self, serde_cbor::Error> {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.data.push(PublishEvent {
+                tags,
+                payload: Payload::compact(event)?,
+            });
+            return Ok(self);
+        }
+        panic!("Calling Publish::event after polling.")
```

Comment:
All these panics need to be visible in the respective method’s doc comment to help people not making the mistake.

Reply:
Done so according to https://rust-lang.github.io/api-guidelines/documentation.html#function-docs-include-error-panic-and-safety-considerations-c-failure